File: Haskell/08-expression-evaluation.hs
import Data.Map.Strict (Map, fromList, (!))


charToFunc :: Map Char (Int -> Int -> Int)
charToFunc = fromList [('+', (+)), ('-', (-))]


splitString :: String -> ([Int], [Char])
splitString string = go string [] [] []
    where
        go :: [Char] -> [Int] -> [Char] -> [Char] -> ([Int], [Char])
        go [] numbers ops currentNumber = (reverse $ (read $ reverse currentNumber) : numbers, reverse ops)
        go (x : xss) numbers ops currentNumber
            | x == '+' || x == '-' = let n = read $ reverse currentNumber in go xss (n : numbers) (x : ops) []
            | otherwise = go xss numbers ops (x : currentNumber)


calculate :: [Int] -> [Char] -> Int
calculate numbers operations =
    foldl (\acc (n, op) -> (charToFunc ! op) acc n) (head numbers) $ zip (tail numbers) operations


main :: IO()
main = do
    string <- getLine
    let (numbers, operations) = splitString string
    let result = calculate numbers operations
    print result

File: Python/08_expression_evaluation.py
import sys
import operator


OPERATIONS = {'+': operator.add, '-': operator.sub}


def split_string(string):
    numbers = []
    operations = []
    current_number = []
    for char in string:
        if char == '+' or char == '-':
            number = int(''.join(current_number))
            current_number = []
            numbers.append(number)
            operations.append(char)
        else:
            current_number.append(char)
    number = int(''.join(current_number))
    numbers.append(number)
    return numbers, operations


def calculate(numbers, operations):
    result = numbers[0]
    for n, op in zip(numbers[1:], operations):
        result = OPERATIONS[op](result, n)
    return result


def main():
    reader = sys.stdin
    string = next(reader).rstrip()
    numbers, operations = split_string(string)
    result = calculate(numbers, operations)
    print(result)


if __name__ == '__main__':
    main()
